runtime/compiler/runtime/StatisticsThread.cpp

```diff
@@ -0,0 +1,197 @@
+   statisticsThread->getStatisticsThreadMonitor()->exit();
+
+   j9thread_set_name(j9thread_self(), "JITaaS Server Statistics Thread");
+
+   TR::CompilationInfo *compInfo = TR::CompilationInfo::get(jitConfig);
+
+   TR::PersistentInfo *persistentInfo = compInfo->getPersistentInfo();
+   persistentInfo->setStartTime(j9time_current_time_millis());
+   persistentInfo->setElapsedTime(0);
+
+   uint64_t oldSyncTime = 0;
+   uint64_t crtTime = 0;
+   uint64_t lastStatistics = 0;
+   uint64_t syncDrift = 0;
+   while(!statisticsThread->getStatisticsThreadExitFlag())
+      {
+      while(!statisticsThread->getStatisticsThreadExitFlag() && j9thread_sleep_interruptable((IDATA) samplingPeriod - syncDrift, 0) == 0)
+         {
+         persistentInfo->updateElapsedTime(samplingPeriod);
+         crtTime += samplingPeriod;
+
+         //adjusting the sampling period window for sync drift.
+         if ((crtTime - oldSyncTime) > (samplingPeriod - syncDrift))
+            {
+            // we need to correct it periodically
+            persistentInfo->setElapsedTime(crtTime);
+            syncDrift = crtTime - oldSyncTime;
+            }
+         else
+            {
+            syncDrift = 0;
+            }
+         oldSyncTime = crtTime;
+         if (persistentInfo->getElapsedTime() - lastStatistics > STATISTICS_LOG_PERIOD)
+            {
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITaaS, "No of clients : 0x%x \n", 0xDEADBEAF);
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITaaS, "Total thread : %d \n", compInfo->getNumUsableCompilationThreads());
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITaaS, "Active Thread : %d \n",compInfo->getNumCompThreadsActive());
+            CpuUtilization *cpuUtil = compInfo->getCpuUtil();
+            if (cpuUtil->isFunctional())
+               {
+
+               TR_VerboseLog::writeLineLocked(TR_Vlog_JITaaS, "CpuLoad %d%% (AvgUsage %d%%) JvmCpu %d%% \n", cpuUtil->getCpuUsage(), cpuUtil->getAvgCpuUsage(), cpuUtil->getVmCpuUsage());
+               }
+            lastStatistics = persistentInfo->getElapsedTime();
+            }
+        }
+
+      // This thread has been interrupted or StatisticsThreadExitFlag flag was set
+      // Sleep just a tiny bit just in case StatisticsThreadExitFlag is set just prior going to sleep
+      //
+      // The elapsed time is now different
+      crtTime = j9time_current_time_millis()-persistentInfo->getStartTime();
+      persistentInfo->setElapsedTime(crtTime);
+      oldSyncTime = crtTime;
+      syncDrift = 0;
+      }
+
+   if (TR::Options::getVerboseOption(TR_VerboseJITaaS))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_JITaaS, "Detaching JITaaSServer listening thread");
+
+   vm->internalVMFunctions->DetachCurrentThread((JavaVM *) vm);
+   statisticsThread->setStatisticsThread(NULL);
+   statisticsThread->getStatisticsThreadMonitor()->enter();
+   statisticsThread->getStatisticsThreadMonitor()->notifyAll();
```

Comment:
Missing statisticsThread->setListenerThreadExitFlag();

Reply:
we will reach here only when the statisticsThreadExitFlag is set from stopStatisticsThread().

Reply:
Correct. The listener does not have a stop routine.